File: sdks/rust/src/realtime.rs
//! Realtime WebSocket client (feature `realtime`).
//!
//! Connects to `GET /realtime/v1/ws/:project` and delivers change / presence
//! frames as an async [`futures_util::Stream`].
//!
//! Protocol source of truth: `crates/basin-realtime/src/ws.rs`
//! (`ClientMsg` / `ServerMsg`) and `presence.rs`. Verified frame shapes:
//!
//! Client → server (`ClientMsg`, `tag = "type"`, snake_case):
//! - `{"type":"subscribe","table":…,"filter"?,"last_event_id"?}`
//! - `{"type":"unsubscribe","table":…}`
//! - `{"type":"presence_track","channel":…,"client_id":…,"metadata":…}`
//! - `{"type":"presence_untrack","channel":…,"client_id":…}`
//! - `{"type":"heartbeat","channel":…,"client_id":…}`
//!
//! Server → client (`ServerMsg`, `tag = "type"`, lowercase):
//! - `{"type":"event",project,table,op,before?,after?,seq}`
//! - `{"type":"subscribed"|"unsubscribed",table}`
//! - `{"type":"error",code,table,missed?}`
//! - `{"type":"gap",table,last_event_id,oldest_in_ring,newest_in_ring}`
//! - `{"type":"presence_state"|"presence_diff",channel,…}`
//! - `{"type":"presenceerror",code,channel,message}` (no underscore — the
//!   server's `PresenceError` variant uses `rename_all = "lowercase"`)
//!
//! Auth: the `Sec-WebSocket-Protocol: basin-v1, <token>` subprotocol form (the
//! server also accepts `Authorization: Bearer`, but the subprotocol form is the
//! one the JS SDK uses and is browser-portable).

use std::collections::HashMap;
use std::time::Duration;

use futures_util::{SinkExt, Stream, StreamExt};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tokio::sync::mpsc;
use tokio_tungstenite::tungstenite::client::IntoClientRequest;
use tokio_tungstenite::tungstenite::http::HeaderValue;
use tokio_tungstenite::tungstenite::Message;

use crate::error::BasinError;
use crate::Client;

// ---------------------------------------------------------------------------
// Wire frame types
// ---------------------------------------------------------------------------

/// A presence channel member.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PresenceEntry {
    /// Client identity (bound to the JWT subject server-side).
    #[serde(default)]
    pub client_id: String,
    /// Arbitrary member metadata.
    #[serde(default)]
    pub metadata: Option<Value>,
    /// RFC 3339 join timestamp, if provided.
    #[serde(default)]
    pub joined_at: Option<String>,
}

/// A server → client frame. Deserialised via the `type` tag.
#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum ServerFrame {
    /// A change event (`op` is `INSERT` / `UPDATE` / `DELETE`).
    Event {
        /// Owning project ULID.
        project: String,
        /// Table the change occurred on.
        table: String,
        /// Change operation: `"INSERT"`, `"UPDATE"`, or `"DELETE"`.
        op: String,
        /// Monotonic sequence number for reconnect resume.
        seq: u64,
        /// Row image before the change (UPDATE / DELETE).
        #[serde(default)]
        before: Option<Value>,
        /// Row image after the change (INSERT / UPDATE).
        #[serde(default)]
        after: Option<Value>,
    },
    /// Acknowledges a successful subscribe.
    Subscribed {
        /// Subscribed table.
        table: String,
    },
    /// Acknowledges a successful unsubscribe.
    Unsubscribed {
        /// Unsubscribed table.
        table: String,
    },
    /// The client lagged and missed events on a channel.
    Error {
        /// Stable error code.
        code: String,
        /// Affected table.
        table: String,
        /// Number of missed events, if known.
        #[serde(default)]
        missed: Option<u64>,
    },
    /// Reconnect-resume gap: the requested `last_event_id` predated the ring;
    /// cold re-sync is needed. The live stream resumes at `newest_in_ring`.
    Gap {
        /// Affected table.
        table: String,
        /// The `last_event_id` the client asked to resume from.
        last_event_id: u64,
        /// Oldest event still in the replay ring.
        oldest_in_ring: u64,
        /// Newest event in the replay ring.
        newest_in_ring: u64,
    },
    /// Initial presence snapshot for a channel.
    #[serde(rename = "presence_state")]
    PresenceState {
        /// Presence channel name.
        channel: String,
        /// Current members.
        #[serde(default)]
        presences: Vec<PresenceEntry>,
    },
    /// Incremental presence delta for a channel.
    #[serde(rename = "presence_diff")]
    PresenceDiff {
        /// Presence channel name.
        channel: String,
        /// Members that joined.
        #[serde(default)]
        joins: Vec<PresenceEntry>,
        /// Members that left.
        #[serde(default)]
        leaves: Vec<PresenceEntry>,
    },
    /// A presence operation was rejected; the connection stays open.
    #[serde(rename = "presenceerror")]
    PresenceError {
        /// Stable error code.
        code: String,
        /// Affected channel.
        channel: String,
        /// Human-readable detail.
        #[serde(default)]
        message: String,
    },
}

impl ServerFrame {
    /// The table or channel this frame pertains to, for routing.
    fn channel_key(&self) -> Option<&str> {
        match self {
            ServerFrame::Event { table, .. }
            | ServerFrame::Subscribed { table }
            | ServerFrame::Unsubscribed { table }
            | ServerFrame::Error { table, .. }
            | ServerFrame::Gap { table, .. } => Some(table),
            ServerFrame::PresenceState { channel, .. }
            | ServerFrame::PresenceDiff { channel, .. }
            | ServerFrame::PresenceError { channel, .. } => Some(channel),
        }
    }
}

/// Options for [`RealtimeClient::listen`].
#[derive(Debug, Clone, Default)]
pub struct SubscribeOptions {
    /// Optional server-side SQL predicate, e.g. `"NEW.status = 'paid'"`.
    pub filter: Option<String>,
    /// Reconnect cursor — the last `seq` the client successfully processed.
    pub last_event_id: Option<u64>,
}

// ---------------------------------------------------------------------------
// Reconnect / backoff
// ---------------------------------------------------------------------------

const BACKOFF_BASE_MS: u64 = 500;
const BACKOFF_MAX_MS: u64 = 30_000;

fn backoff(attempt: u32) -> Duration {
    let ms = BACKOFF_BASE_MS.saturating_mul(2u64.saturating_pow(attempt));
    Duration::from_millis(ms.min(BACKOFF_MAX_MS))
}

// ---------------------------------------------------------------------------
// RealtimeClient
// ---------------------------------------------------------------------------

/// Async WebSocket client for `GET /realtime/v1/ws/:project`.
///
/// Obtain via [`Client::realtime`](crate::Client::realtime). The primary API is
/// [`listen`](Self::listen), which returns a [`Stream`] of [`ServerFrame`]s for
/// one table or presence channel and transparently reconnects with exponential
/// backoff, re-issuing the subscription on each reconnect.
#[derive(Clone)]
pub struct RealtimeClient {
    client: Client,
}

impl RealtimeClient {
    pub(crate) fn new(client: Client) -> Self {
        Self { client }
    }

    async fn connect_url(&self) -> Result<(String, String), BasinError> {
        let project = self.client.resolve_project_id().await.ok_or_else(|| {
            BasinError::invalid_request(
                "realtime requires a project id — set project_id on the client builder",
            )
        })?;
        let token = self
            .client
            .transport()
            .bearer()
            .await?
            .ok_or_else(|| {
                BasinError::Realtime("realtime requires a JWT or API key".to_string())
            })?;
        let ws_base = self
            .client
            .base_url()
            .replacen("https://", "wss://", 1)
            .replacen("http://", "ws://", 1);
        let url = format!("{ws_base}/realtime/v1/ws/{project}");
        Ok((url, token))
    }

    /// Subscribe to `table` and stream change frames.
    ///
    /// The returned [`Stream`] yields [`ServerFrame`]s (events, errors, gaps).
    /// On an unexpected disconnect the client reconnects with exponential
    /// backoff (0.5 s → 1 s → 2 s … capped at 30 s) and re-issues the
    /// subscription. Pass [`SubscribeOptions::last_event_id`] for server-side
    /// replay of missed events. Dropping the stream tears down the connection.
    pub fn listen(
        &self,
        table: impl Into<String>,
        opts: SubscribeOptions,
    ) -> impl Stream<Item = Result<ServerFrame, BasinError>> {
        let table = table.into();
        let this = self.clone();
        let (tx, rx) = mpsc::unbounded_channel();

        tokio::spawn(async move {
            let mut attempt: u32 = 0;
            loop {
                match this.run_session(&table, &opts, &tx).await {
                    // Graceful end (receiver dropped) — stop.
                    Ok(SessionEnd::ReceiverGone) => break,
                    // Connection dropped — reconnect after backoff.
                    Ok(SessionEnd::Disconnected) => {}
                    Err(e) => {
                        // Surface the error to the consumer; then retry.
                        if tx.send(Err(e)).is_err() {
                            break;
                        }
                    }
                }
                attempt = attempt.saturating_add(1);
                tokio::time::sleep(backoff(attempt - 1)).await;
                if tx.is_closed() {
                    break;
                }
            }
        });

        tokio_stream(rx)
    }

    /// Run a single WebSocket session: connect, subscribe, pump frames until
    /// the socket closes or the receiver is dropped.
    async fn run_session(
        &self,
        table: &str,
        opts: &SubscribeOptions,
        tx: &mpsc::UnboundedSender<Result<ServerFrame, BasinError>>,
    ) -> Result<SessionEnd, BasinError> {
        let (url, token) = self.connect_url().await?;

        let mut request = url
            .into_client_request()
            .map_err(|e| BasinError::Realtime(e.to_string()))?;
        let proto = format!("basin-v1, {token}");
        request.headers_mut().insert(
            "Sec-WebSocket-Protocol",
            HeaderValue::from_str(&proto).map_err(|e| BasinError::Realtime(e.to_string()))?,
        );

        let (ws, _resp) = tokio_tungstenite::connect_async(request)
            .await
            .map_err(|e| BasinError::Realtime(e.to_string()))?;
        let (mut sink, mut stream) = ws.split();

        // Send the subscribe frame.
        let mut sub = serde_json::Map::new();
        sub.insert("type".into(), Value::String("subscribe".into()));
        sub.insert("table".into(), Value::String(table.to_string()));
        if let Some(f) = &opts.filter {
            sub.insert("filter".into(), Value::String(f.clone()));
        }
        if let Some(id) = opts.last_event_id {
            sub.insert("last_event_id".into(), Value::from(id));
        }
        sink.send(Message::Text(Value::Object(sub).to_string()))
            .await
            .map_err(|e| BasinError::Realtime(e.to_string()))?;

        while let Some(msg) = stream.next().await {
            let msg = match msg {
                Ok(m) => m,
                Err(_) => return Ok(SessionEnd::Disconnected),
            };
            match msg {
                Message::Text(text) => {
                    if let Some(frame) = parse_frame(&text) {
                        // Route only frames for this table/channel; the server
                        // multiplexes one socket but a listener watches one key.
                        let relevant = frame
                            .channel_key()
                            .map(|k| k == table)
                            .unwrap_or(true);
                        if relevant && tx.send(Ok(frame)).is_err() {
                            return Ok(SessionEnd::ReceiverGone);
                        }
                    }
                }
                Message::Ping(p) => {
                    let _ = sink.send(Message::Pong(p)).await;
                }
                Message::Close(_) => return Ok(SessionEnd::Disconnected),
                _ => {}
            }
            if tx.is_closed() {
                return Ok(SessionEnd::ReceiverGone);
            }
        }
        Ok(SessionEnd::Disconnected)
    }
}

enum SessionEnd {
    Disconnected,
    ReceiverGone,
}

/// Parse a raw text frame into a [`ServerFrame`], returning `None` for unknown
/// or malformed frames (the caller logs and skips, matching the Python SDK).
fn parse_frame(raw: &str) -> Option<ServerFrame> {
    serde_json::from_str::<ServerFrame>(raw).ok()
}

/// Wrap an unbounded receiver as a [`Stream`].
fn tokio_stream<T>(mut rx: mpsc::UnboundedReceiver<T>) -> impl Stream<Item = T> {
    futures_util::stream::poll_fn(move |cx| rx.poll_recv(cx))
}

/// A typed presence metadata map for [`RealtimeClient`] presence helpers.
pub type PresenceMetadata = HashMap<String, Value>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_event_frame() {
        let raw = r#"{"type":"event","project":"p","table":"orders","op":"INSERT","seq":7,"after":{"id":1}}"#;
        let f = parse_frame(raw).unwrap();
        match f {
            ServerFrame::Event {
                table, op, seq, after, ..
            } => {
                assert_eq!(table, "orders");
                assert_eq!(op, "INSERT");
                assert_eq!(seq, 7);
                assert_eq!(after.unwrap()["id"], 1);
            }
            _ => panic!("wrong variant"),
        }
    }

    #[test]
    fn parses_gap_frame() {
        let raw = r#"{"type":"gap","table":"t","last_event_id":3,"oldest_in_ring":10,"newest_in_ring":20}"#;
        assert!(matches!(parse_frame(raw), Some(ServerFrame::Gap { .. })));
    }

    #[test]
    fn parses_presenceerror_no_underscore() {
        let raw = r#"{"type":"presenceerror","code":"impersonation","channel":"room:1","message":"nope"}"#;
        assert!(matches!(
            parse_frame(raw),
            Some(ServerFrame::PresenceError { .. })
        ));
    }

    #[test]
    fn parses_presence_state() {
        let raw = r#"{"type":"presence_state","channel":"room:1","presences":[{"client_id":"c1"}]}"#;
        match parse_frame(raw).unwrap() {
            ServerFrame::PresenceState { channel, presences } => {
                assert_eq!(channel, "room:1");
                assert_eq!(presences.len(), 1);
                assert_eq!(presences[0].client_id, "c1");
            }
            _ => panic!("wrong variant"),
        }
    }

    #[test]
    fn unknown_frame_is_none() {
        assert!(parse_frame(r#"{"type":"weird"}"#).is_none());
        assert!(parse_frame("not json").is_none());
    }

    #[test]
    fn backoff_caps_at_30s() {
        assert_eq!(backoff(0), Duration::from_millis(500));
        assert_eq!(backoff(1), Duration::from_millis(1000));
        assert_eq!(backoff(100), Duration::from_millis(30_000));
    }

    #[test]
    fn channel_key_routes_event_to_table() {
        let f = ServerFrame::Event {
            project: "p".into(),
            table: "orders".into(),
            op: "INSERT".into(),
            seq: 1,
            before: None,
            after: None,
        };
        assert_eq!(f.channel_key(), Some("orders"));
    }
}

File: sdks/dart/lib/src/realtime.dart
/// Realtime WebSocket client for Basin.
///
/// Connects to GET /realtime/v1/ws/:project and delivers change events as
/// Dart Streams.
///
/// Protocol source of truth: crates/basin-realtime/src/ws.rs
/// (ClientMsg / ServerMsg + presence.rs helpers).
///
/// Auth: Sec-WebSocket-Protocol: basin-v1, <token> subprotocol header.
/// The server also accepts Authorization: Bearer but browsers and standard
/// WebSocket implementations make the subprotocol form simpler for
/// client-originated connections.
///
/// # Flutter Web vs. native
/// web_socket_channel is cross-platform. On Flutter Web the underlying
/// implementation is dart:html WebSocket; on native (iOS, Android, desktop,
/// server) it uses dart:io WebSocket. Both accept the subprotocol list.
/// The subprotocol auth form is therefore the correct choice for browser
/// compatibility (browsers cannot set custom HTTP headers on WS upgrades).
///
/// # Reconnect behaviour
/// On unexpected disconnect, the client reconnects with exponential backoff
/// (0.5s → 1s → 2s … capped at 30s) and re-issues all active subscriptions.
/// Pass [lastEventId] to [listen] / [subscribe] for server-side replay.
library;

import 'dart:async';
import 'dart:convert';
import 'dart:math' as math;

import 'package:web_socket_channel/web_socket_channel.dart';

import 'errors.dart';
import 'types.dart';

const double _backoffBase = 0.5;
const double _backoffMax = 30.0;
const double _backoffFactor = 2.0;

Duration _backoff(int attempt) {
  final secs = _backoffBase * math.pow(_backoffFactor, attempt);
  return Duration(
    milliseconds: (math.min(secs, _backoffMax) * 1000).round(),
  );
}

/// Parse a raw JSON text frame from the server into a typed object.
/// Returns null for unknown or malformed frames.
Object? _parseFrame(String raw) {
  Map<String, dynamic> d;
  try {
    d = jsonDecode(raw) as Map<String, dynamic>;
  } catch (_) {
    return null;
  }
  final type = d['type'] as String?;
  if (type == null) return null;
  return switch (type) {
    'event' => RealtimeEvent.fromJson(d),
    'error' => RealtimeErrorFrame.fromJson(d),
    'gap' => RealtimeGapFrame.fromJson(d),
    'subscribed' => _SubscribedFrame(d['table'] as String? ?? ''),
    'unsubscribed' => _UnsubscribedFrame(d['table'] as String? ?? ''),
    'presence_state' => PresenceStateFrame.fromJson(d),
    'presence_diff' => PresenceDiffFrame.fromJson(d),
    // "presenceerror" — no underscore; server uses rename_all="lowercase"
    'presenceerror' => PresenceErrorFrame.fromJson(d),
    _ => null,
  };
}

class _SubscribedFrame {
  final String table;
  const _SubscribedFrame(this.table);
}

class _UnsubscribedFrame {
  final String table;
  const _UnsubscribedFrame(this.table);
}

/// Handle returned by [RealtimeClient.subscribe] (callback-based API).
class ChannelHandle {
  final String table;
  final StreamSubscription<void> _subscription;
  final void Function() _cancel;

  ChannelHandle._(this.table, this._subscription, this._cancel);

  /// Stop receiving events and unsubscribe from the server.
  Future<void> unsubscribe() async {
    _cancel();
    await _subscription.cancel();
  }
}

/// WebSocket client for GET /realtime/v1/ws/:project.
///
/// Obtain via [BasinClient.realtime]; do not construct directly.
class RealtimeClient {
  final String _baseUrl;
  final Future<String?> Function() _getToken;
  final String? Function() _projectId;

  /// Injectable WebSocket factory (for tests). Defaults to
  /// [WebSocketChannel.connect] from package:web_socket_channel.
  final WebSocketChannel Function(Uri, {Iterable<String>? protocols})?
  _wsFactory;

  WebSocketChannel? _ws;
  StreamSubscription<dynamic>? _recvSub;
  bool _closed = false;

  // Subscriptions: table → options
  final Map<String, Map<String, dynamic>> _subscriptions = {};

  // Per-table stream controllers
  final Map<String, List<StreamController<Object>>> _controllers = {};

  // Pending acks
  final Map<String, Completer<void>> _pendingSubscribed = {};
  final Map<String, Completer<void>> _pendingUnsubscribed = {};

  // Reconnect lock
  bool _reconnecting = false;

  RealtimeClient({
    required String baseUrl,
    required Future<String?> Function() getToken,
    required String? Function() projectId,
    WebSocketChannel Function(Uri, {Iterable<String>? protocols})? wsFactory,
  }) : _baseUrl = baseUrl,
       _getToken = getToken,
       _projectId = projectId,
       _wsFactory = wsFactory;

  bool get isConnected => _ws != null;

  // ------------------------------------------------------------------
  // Connection management
  // ------------------------------------------------------------------

  /// Open the WebSocket connection. No-op if already connected.
  Future<void> connect() async {
    if (_ws != null) return;
    await _doConnect();
  }

  Future<void> _doConnect() async {
    final project = _projectId();
    if (project == null) {
      throw const BasinApiError(
        'E_INVALID_REQUEST',
        'realtime requires a project id — pass projectId to BasinClient.create()',
        0,
      );
    }
    final token = await _getToken();
    if (token == null) {
      throw const BasinApiError(
        'E_UNAUTHENTICATED',
        'realtime requires a JWT or API key',
        0,
      );
    }

    final wsBase = _baseUrl
        .replaceAll(RegExp(r'/$'), '')
        .replaceFirst('https://', 'wss://')
        .replaceFirst('http://', 'ws://');
    final uri = Uri.parse('$wsBase/realtime/v1/ws/$project');

    final protocols = ['basin-v1', token];

    final factory = _wsFactory ?? WebSocketChannel.connect;
    _ws = factory(uri, protocols: protocols);

    _recvSub = _ws!.stream.listen(
      _onData,
      onError: _onError,
      onDone: _onDone,
      cancelOnError: false,
    );
  }

  void _onData(dynamic raw) {
    if (raw is! String) return;
    final frame = _parseFrame(raw);
    if (frame == null) return;
    _dispatch(frame);
  }

  void _onError(Object err, StackTrace st) {
    // Will also trigger _onDone.
  }

  void _onDone() {
    _ws = null;
    if (!_closed) {
      unawaited(_reconnect());
    } else {
      _drainAll(null);
    }
  }

  void _dispatch(Object frame) {
    if (frame is RealtimeEvent) {
      _broadcastToTable(frame.table, frame);
    } else if (frame is RealtimeErrorFrame) {
      _broadcastToTable(frame.table, frame);
    } else if (frame is RealtimeGapFrame) {
      _broadcastToTable(frame.table, frame);
    } else if (frame is _SubscribedFrame) {
      final c = _pendingSubscribed.remove(frame.table);
      if (c != null && !c.isCompleted) c.complete();
    } else if (frame is _UnsubscribedFrame) {
      final c = _pendingUnsubscribed.remove(frame.table);
      if (c != null && !c.isCompleted) c.complete();
    } else if (frame is PresenceStateFrame) {
      _broadcastToTable(frame.channel, frame);
    } else if (frame is PresenceDiffFrame) {
      _broadcastToTable(frame.channel, frame);
    } else if (frame is PresenceErrorFrame) {
      _broadcastToTable(frame.channel, frame);
    }
  }

  void _broadcastToTable(String key, Object event) {
    for (final ctrl in _controllers[key] ?? <StreamController<Object>>[]) {
      if (!ctrl.isClosed) ctrl.add(event);
    }
  }

  void _drainAll(Object? sentinel) {
    for (final ctrls in _controllers.values) {
      for (final ctrl in ctrls) {
        if (!ctrl.isClosed) ctrl.close();
      }
    }
  }

  /// Close the WebSocket connection and all active streams.
  Future<void> disconnect() async {
    _closed = true;
    await _recvSub?.cancel();
    _recvSub = null;
    await _ws?.sink.close();
    _ws = null;
    _drainAll(null);
  }

  // ------------------------------------------------------------------
  // Reconnect
  // ------------------------------------------------------------------

  Future<void> _reconnect() async {
    if (_reconnecting) return;
    _reconnecting = true;
    var attempt = 0;
    while (!_closed) {
      await Future<void>.delayed(_backoff(attempt));
      attempt++;
      try {
        await _doConnect();
        await _resubscribe();
        _reconnecting = false;
        return;
      } catch (_) {
        // try again
      }
    }
    _reconnecting = false;
  }

  Future<void> _resubscribe() async {
    for (final entry in _subscriptions.entries) {
      try {
        await _sendSubscribe(entry.key, entry.value);
        final c = Completer<void>();
        _pendingSubscribed[entry.key] = c;
        await c.future.timeout(const Duration(seconds: 10));
      } catch (_) {}
    }
  }

  // ------------------------------------------------------------------
  // Subscribe / unsubscribe
  // ------------------------------------------------------------------

  void _send(Map<String, dynamic> msg) {
    if (_ws == null) {
      throw const BasinApiError('E_UNKNOWN', 'realtime socket is not open', 0);
    }
    _ws!.sink.add(jsonEncode(msg));
  }

  Future<void> _sendSubscribe(
    String table,
    Map<String, dynamic> opts,
  ) async {
    final msg = <String, dynamic>{'type': 'subscribe', 'table': table};
    if (opts['filter'] != null) msg['filter'] = opts['filter'];
    if (opts['last_event_id'] != null) {
      msg['last_event_id'] = opts['last_event_id'];
    }
    _send(msg);
  }

  Future<void> _subscribe(
    String table,
    Map<String, dynamic> opts,
  ) async {
    await connect();
    _subscriptions[table] = opts;
    _controllers.putIfAbsent(table, () => []);

    final c = Completer<void>();
    _pendingSubscribed[table] = c;
    await _sendSubscribe(table, opts);
    await c.future.timeout(const Duration(seconds: 30));
  }

  Future<void> _unsubscribe(String table) async {
    if (_ws == null) return;
    final c = Completer<void>();
    _pendingUnsubscribed[table] = c;
    _send({'type': 'unsubscribe', 'table': table});
    try {
      await c.future.timeout(const Duration(seconds: 10));
    } on TimeoutException {
      // Server may not ack; proceed with local cleanup.
    }
    _subscriptions.remove(table);
  }

  // ------------------------------------------------------------------
  // Public API — Stream (primary)
  // ------------------------------------------------------------------

  /// Stream of change events for [table].
  ///
  /// Subscribes on first listen and unsubscribes when the subscription is
  /// cancelled (if this is the last listener for that table).
  ///
  /// [filter]: optional server-side SQL predicate, e.g. "NEW.status = 'paid'".
  /// [lastEventId]: reconnect cursor — last [RealtimeEvent.seq] processed;
  ///   server replays missed events from its ring, or emits a
  ///   [RealtimeGapFrame] if the ring has been evicted.
  ///
  /// Emits [RealtimeEvent], [RealtimeErrorFrame], and [RealtimeGapFrame].
  Stream<Object> listen(
    String table, {
    String? filter,
    int? lastEventId,
  }) {
    late StreamController<Object> ctrl;
    ctrl = StreamController<Object>(
      onListen: () async {
        _controllers.putIfAbsent(table, () => []).add(ctrl);
        try {
          await _subscribe(table, {
            if (filter != null) 'filter': filter,
            if (lastEventId != null) 'last_event_id': lastEventId,
          });
        } catch (e) {
          ctrl.addError(e);
        }
      },
      onCancel: () async {
        final ctrls = _controllers[table];
        ctrls?.remove(ctrl);
        if ((ctrls?.isEmpty ?? true) && _subscriptions.containsKey(table)) {
          await _unsubscribe(table);
          _controllers.remove(table);
        }
      },
    );
    return ctrl.stream;
  }

  // ------------------------------------------------------------------
  // Public API — callback-based (adapter over listen())
  // ------------------------------------------------------------------

  /// Subscribe to [table] with a callback.
  ///
  /// Returns a [ChannelHandle] whose [ChannelHandle.unsubscribe] stops the
  /// subscription.
  Future<ChannelHandle> subscribe(
    String table,
    void Function(Object event) onEvent, {
    String? filter,
    int? lastEventId,
  }) async {
    final stream = listen(table, filter: filter, lastEventId: lastEventId);
    late StreamSubscription<void> sub;
    late ChannelHandle handle;

    sub = stream.listen(
      onEvent,
      onError: (Object _) {},
      cancelOnError: false,
    );

    handle = ChannelHandle._(table, sub, () => sub.cancel());
    return handle;
  }

  // ------------------------------------------------------------------
  // Presence API
  // ------------------------------------------------------------------

  /// Send {"type":"presence_track",...} to register in a presence channel.
  ///
  /// The server binds client_id to the JWT identity — a mismatch is rejected
  /// (6.SEC.P1) and a presenceerror frame is delivered.
  Future<void> presenceTrack(
    String channel,
    String clientId, {
    dynamic metadata,
  }) async {
    await connect();
    _send({
      'type': 'presence_track',
      'channel': channel,
      'client_id': clientId,
      'metadata': metadata ?? {},
    });
  }

  /// Send {"type":"presence_untrack",...}.
  Future<void> presenceUntrack(String channel, String clientId) async {
    await connect();
    _send({
      'type': 'presence_untrack',
      'channel': channel,
      'client_id': clientId,
    });
  }

  /// Send {"type":"heartbeat",...} — refresh the presence TTL.
  Future<void> presenceHeartbeat(String channel, String clientId) async {
    await connect();
    _send({'type': 'heartbeat', 'channel': channel, 'client_id': clientId});
  }

  /// Stream of presence frames (presence_state, presence_diff, presenceerror)
  /// for [channel].
  ///
  /// Frames arrive after [presenceTrack] registers the client.
  Stream<Object> listenPresence(String channel) {
    late StreamController<Object> ctrl;
    ctrl = StreamController<Object>(
      onListen: () {
        _controllers.putIfAbsent(channel, () => []).add(ctrl);
      },
      onCancel: () {
        final ctrls = _controllers[channel];
        ctrls?.remove(ctrl);
        if (ctrls?.isEmpty ?? true) _controllers.remove(channel);
      },
    );
    return ctrl.stream;
  }
}

// Suppress unawaited_futures lint for fire-and-forget reconnect calls.
void unawaited(Future<void> f) {}
